### PR TITLE
style: implement shellcheck autofix in `brew style --fix`

### PR DESCRIPTION
**Pull request by XuehaiPan**
- [X] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [X] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [X] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [ ] Have you successfully run `brew style` with your changes locally?
- [ ] Have you successfully run `brew typecheck` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?

-----

Improve Shellcheck usage, see #11086 and https://github.com/Homebrew/brew/issues/11086#issuecomment-946451748.

### CODE REVIEW

File: Library/Homebrew/style.rb [hunk 1]
```diff
@@ -164,17 +164,24 @@ def run_rubocop(files, output_type,
...
     end
 
-    def run_shellcheck(files, output_type)
+    def run_shellcheck(files, output_type, fix: false)
       files = shell_scripts if files.blank?
 
-      args = ["--shell=bash", "--enable=all", "--external-sources", "--source-path=#{HOMEBREW_LIBRARY}", "--", *files]
+      files = files.map(&:realpath)
+
+      args = ["--shell=bash", "--enable=all", "--external-sources", "--source-path=#{HOMEBREW_LIBRARY}", "--"]
```
Comment (XuehaiPan) [inline Library/Homebrew/style.rb:172]: Should I add `"--exclude=SC2310,SC2311,SC2312"` option here or add `# shellcheck=...` in individual file? There are too many files involved.

https://github.com/koalaman/shellcheck/wiki/SC2310
https://github.com/koalaman/shellcheck/wiki/SC2311
https://github.com/koalaman/shellcheck/wiki/SC2312

File: Library/Homebrew/style.rb [hunk 2]
```diff
@@ -164,17 +164,24 @@ def run_rubocop(files, output_type,
...
+
+      if fix
+        patch = system_command shellcheck, args: ["--format=diff", *args, *files]
+        system_command "patch", args: ["-d", "/", "-p0"], input: patch.stdout
+      end
```
Comment (XuehaiPan) [inline Library/Homebrew/style.rb:177]: Is there a better way to apply patches? This system call requires a `patch` command. The patch file generated by `shellcheck` uses absolute paths which cannot apply by `git apply`.

File: Library/Homebrew/style.rb (hunk 1 above)
Comment (Bo98) [reply Library/Homebrew/style.rb:172]: As a temporary fix for CI (I need to get other PRs unblocked), I've applied this exclusion in #12393.

File: Library/Homebrew/style.rb [hunk 3]
```diff
@@ -178,6 +180,12 @@ def run_shellcheck(files, output_type)
         *files,
       ]
 
+      if fix
+        patches = system_command(shellcheck, args: ["--format=diff", *args]).stdout
+        patch_command = %w[patch -g 0 -f -d / -p0]
```
Comment (MikeMcQuaid) [inline Library/Homebrew/style.rb:185]: Can you:
- use long flags where possible
- add a comment to explain why this is needed and what these flags are doing

Thanks!

**Review by MikeMcQuaid — COMMENTED** [review]
Looking great so far!

**Review by MikeMcQuaid — APPROVED** [review]

File: Library/Homebrew/style.rb [hunk 4]
```diff
@@ -178,6 +180,17 @@ def run_shellcheck(files, output_type)
...
 
+      if fix
+        # patch options:
+        #   --get=0       : suppress environment variable `PATCH_GET`, ignore RCS, ClearCase, Perforce, and SCCS
+        #   --force       : we know what we are doing, force apply patches
+        #   --directory=/ : change to root directory, since we use absolute file paths
+        #   --strip=0     : do not strip path prefixes, since we are at root directory
+        patch_command = %w[patch --get=0 --force --directory=/ --strip=0]
+        patches = system_command(shellcheck, args: ["--format=diff", *args]).stdout
+        Utils.popen_write(*patch_command) { |p| p.write(patches) }
```
Comment (MikeMcQuaid) [inline Library/Homebrew/style.rb:191]:
```suggestion
        Utils.safe_popen_write(*patch_command) { |p| p.write(patches) }
```

Probably as we don't want to ignore failures, right?

**Review by MikeMcQuaid — APPROVED** [review]

**Comment by MikeMcQuaid** [comment]
Thanks again @XuehaiPan!

**Pull request merged**